File: tests/test_user_quantiles.py
import numpy as np

from gbrs import GBRS


def test_user_quantiles_python():
    print("\n=== Testing GBRS User-Defined Quantiles (Python) ===")

    # Generate synthetic data
    np.random.seed(42)
    n = 200
    X = np.random.rand(n, 3)
    # Feature 0: [0, 1]
    # Feature 1: [0, 1]
    # Feature 2: [0, 1]

    # Target depends on X[0] > 0.5 and X[1] > 0.7
    y = (
        (X[:, 0] > 0.5).astype(float)
        + (X[:, 1] > 0.7).astype(float)
        + np.random.normal(0, 0.1, n)
    )

    # Define custom quantiles
    # Must be a list of numpy arrays, one for each feature
    # If a feature should use auto-quantiles, pass None (or handle in wrapper if implemented)
    # Here we define for all
    user_quantiles = [
        np.array([0.2, 0.5, 0.8]),  # Feature 0
        np.array([0.3, 0.7]),  # Feature 1
        np.array([0.5]),  # Feature 2
    ]

    print("Fitting model with user_quantiles...")
    model = GBRS(n_iter=50, lr=0.1, n_quantiles=10)
    model.fit(X, y, user_quantiles=user_quantiles)

    print("Model fitted successfully.")

    # Verify splits used

    idxs = model._model.get_idxs()
    split_vals = model._model.get_split_val()

    # Check Feature 0 splits
    f0_splits = split_vals[idxs == 0]
    print(f"Splits for Feature 0: {np.unique(f0_splits)}")

    # We expect splits to be from our user_quantiles (approximate due to float precision)
    # Note: The model might not use ALL provided quantiles if they don't improve the loss

    expected_f0 = [0.2, 0.5, 0.8]
    found_f0 = False
    for s in f0_splits:
        if any(np.isclose(s, expected_f0, atol=1e-5)):
            found_f0 = True
            break

    if found_f0:
        print("✓ Found expected custom split for Feature 0")
    else:
        print(
            "⚠ No custom splits for Feature 0 were selected (possible if not informative)"
        )

    # Check Feature 1 splits
    f1_splits = split_vals[idxs == 1]
    print(f"Splits for Feature 1: {np.unique(f1_splits)}")

    expected_f1 = [0.3, 0.7]
    found_f1 = False
    for s in f1_splits:
        if any(np.isclose(s, expected_f1, atol=1e-5)):
            found_f1 = True
            break

    if found_f1:
        print("✓ Found expected custom split for Feature 1")

    # Test print
    print("\n=== Model Print ===")
    model.print()


if __name__ == "__main__":
    test_user_quantiles_python()

File: tests/test_user_quantiles.R
library(gbrs)
library(survival)

# Load a real survival dataset
data(veteran)

cat("=== Testing GBRS User-Defined Quantiles (R) ===\n\n")

# Define custom quantiles for selected features
# trt is binary (1, 2)
# age is continuous
custom_q <- list(
  trt = c(1.5),             # Split at 1.5
  age = c(40, 60, 80)       # Custom age bins
)

# Fit GBRS survival model with user_quantiles
cat("Fitting GBRS survival model with custom quantiles...\n")
model <- gbrs(
  Surv(time, status) ~ trt + age + celltype,
  data = veteran,
  n_max = 50,
  lr = 0.05,
  n_quantiles = 10,

  objective = "survival",
  user_quantiles = custom_q
)

cat("Model fitted successfully!\n\n")

# Verify that the model used the custom quantiles
# We check the split values in the model weights
weights <- model$weights
print(head(weights))

# Check trt splits
trt_idx <- which(attr(terms(model$formula), "term.labels") == "trt") - 1
trt_splits <- weights$split_val[weights$idx == trt_idx]
cat("\nSplits for 'trt' (expected ~1.5):", paste(sort(unique(trt_splits)), collapse=", "), "\n")

if (any(abs(trt_splits - 1.5) < 1e-5)) {
    cat("✓ Custom split for 'trt' found.\n")
} else {
    cat("✗ Custom split for 'trt' NOT found.\n")
}

# Check age splits
age_idx <- which(attr(terms(model$formula), "term.labels") == "age") - 1
age_splits <- weights$split_val[weights$idx == age_idx]
cat("Splits for 'age' (expected subset of 40, 60, 80):", paste(sort(unique(age_splits)), collapse=", "), "\n")

expected_age_splits <- c(40, 60, 80)
found_splits <- intersect(round(age_splits), expected_age_splits)
if (length(found_splits) > 0) {
    cat("✓ Found custom splits for 'age':", paste(found_splits, collapse=", "), "\n")
} else {
    cat("⚠ No custom splits for 'age' were selected by the model (this can happen if they are not informative).\n")
}

# Test standard print (horizontal default)
cat("\n=== Standard Print (Horizontal) ===\n")
print(model)

# Test legacy vertical print
cat("\n=== Legacy Vertical Print ===\n")
print.vertical(model)

# Test LaTeX print (vertical)
cat("\n=== LaTeX Print (Vertical) ===\n")
print(model, format="latex")

# Test Markdown print (vertical)
cat("\n=== Markdown Print (Vertical) ===\n")
print(model, format="md")

cat("\nTest completed.\n")
